### PR TITLE
rustdoc: Preprocess intra-doc links consistently between crate loader and link resolver

### LINKED ISSUE #84046: Rustdoc ICE: Referencing a workspace crate's root in an intra-doc link triggers an ICE
### Code

The MCVE includes multiple crates so I've [put it on a GitHub repo](https://github.com/ThePuzzlemaker/rustdoc-ice-mcve).

The command used for Rustdoc is just `cargo +nightly doc`.

### Meta

`rustc --version --verbose`:
```
rustc 1.53.0-nightly (dae9d6ac3 2021-04-09)
binary: rustc
commit-hash: dae9d6ac3e9a8fb389cfe471b30f4d72c3122196
commit-date: 2021-04-09
host: x86_64-unknown-linux-gnu
release: 1.53.0-nightly
LLVM version: 12.0.0
```

### Error output

```
 Documenting rustdoc_ice_mcve v0.1.0 (/home/USER/Code/Rust/foo)
thread 'rustc' panicked at 'index out of bounds: the len is 18 but the index is 18', compiler/rustc_metadata/src/creader.rs:137:21
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

error: internal compiler error: unexpected panic

error: Unrecognized option: 'crate-version'

error: could not document `rustdoc_ice_mcve`

Caused by:
  process didn't exit successfully: `rustdoc --edition=2018 --crate-type lib --crate-name rustdoc_ice_mcve src/lib.rs -o /home/USER/Code/Rust/foo/target/doc --error-format=json --json=diagnostic-rendered-ansi -L dependency=/home/USER/Code/Rust/foo/target/debug/deps --extern workspace_crate=/home/USER/Code/Rust/foo/target/debug/deps/libworkspace_crate-46c87e60a6b67395.rmeta --crate-version 0.1.0` (exit status: 1)
```
<details><summary><strong>Backtrace</strong></summary>
<p>
I've included the backtrace with `--verbose` as it mentioned `--crate-version`.

```
       Fresh workspace_crate v0.1.0 (/home/USER/Code/Rust/foo/workspace_crate)
 Documenting rustdoc_ice_mcve v0.1.0 (/home/USER/Code/Rust/foo)
     Running `rustdoc --edition=2018 --crate-type lib --crate-name rustdoc_ice_mcve src/lib.rs -o /home/USER/Code/Rust/foo/target/doc --error-format=json --json=diagnostic-rendered-ansi -L dependency=/home/USER/Code/Rust/foo/target/debug/deps --extern workspace_crate=/home/USER/Code/Rust/foo/target/debug/deps/libworkspace_crate-46c87e60a6b67395.rmeta --crate-version 0.1.0`
thread 'rustc' panicked at 'index out of bounds: the len is 18 but the index is 18', compiler/rustc_metadata/src/creader.rs:137:21
stack backtrace:
   0: rust_begin_unwind
             at /rustc/dae9d6ac3e9a8fb389cfe471b30f4d72c3122196/library/std/src/panicking.rs:493:5
   1: core::panicking::panic_fmt
             at /rustc/dae9d6ac3e9a8fb389cfe471b30f4d72c3122196/library/core/src/panicking.rs:92:14
   2: core::panicking::panic_bounds_check
             at /rustc/dae9d6ac3e9a8fb389cfe471b30f4d72c3122196/library/core/src/panicking.rs:69:5
   3: rustc_metadata::rmeta::decoder::cstore_impl::provide_extern::crate_name
   4: rustc_query_system::dep_graph::graph::DepGraph<K>::with_task_impl
   5: rustc_data_structures::stack::ensure_sufficient_stack
   6: rustc_query_system::query::plumbing::force_query_with_job
   7: rustc_query_system::query::plumbing::get_query_impl
   8: <rustc_query_impl::Queries as rustc_middle::ty::query::QueryEngine>::crate_name
   9: rustdoc::clean::inline::record_extern_fqn
  10: rustdoc::clean::utils::register_res
  11: rustdoc::passes::collect_intra_doc_links::LinkCollector::resolve_link
  12: <rustdoc::passes::collect_intra_doc_links::LinkCollector as rustdoc::fold::DocFolder>::fold_item
  13: alloc::vec::source_iter_marker::<impl alloc::vec::spec_from_iter::SpecFromIter<T,I> for alloc::vec::Vec<T>>::from_iter
  14: rustdoc::fold::DocFolder::fold_inner_recur
  15: rustdoc::fold::DocFolder::fold_item_recur
  16: <rustdoc::passes::collect_intra_doc_links::LinkCollector as rustdoc::fold::DocFolder>::fold_item
  17: rustdoc::passes::collect_intra_doc_links::collect_intra_doc_links
  18: rustdoc::core::run_global_ctxt
  19: rustc_interface::passes::QueryContext::enter
  20: rustc_interface::queries::<impl rustc_interface::interface::Compiler>::enter
  21: rustc_span::with_source_map
  22: rustc_interface::interface::create_compiler_and_run
  23: rustc_span::with_session_globals
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.

error: internal compiler error: unexpected panic

error: Unrecognized option: 'crate-version'

error: could not document `rustdoc_ice_mcve`

Caused by:
  process didn't exit successfully: `rustdoc --edition=2018 --crate-type lib --crate-name rustdoc_ice_mcve src/lib.rs -o /home/USER/Code/Rust/foo/target/doc --error-format=json --json=diagnostic-rendered-ansi -L dependency=/home/USER/Code/Rust/foo/target/debug/deps --extern workspace_crate=/home/USER/Code/Rust/foo/target/debug/deps/libworkspace_crate-46c87e60a6b67395.rmeta --crate-version 0.1.0` (exit status: 1)
```

</p>
</details>

### PR DESCRIPTION
**Pull request by jyn514**
Previously, rustdoc would panic on links to external crates if they were
surrounded by backticks or the first path segment contained generics.

This is more extensive than the fix I suggested in #84046; it makes the preprocessing exactly the same rather than trying to duplicate only the necessary logic. This makes fixing the generics a lot easier as it turns out.

Fixes https://github.com/rust-lang/rust/issues/84046.

r? @Manishearth

### CODE REVIEW

**Comment by jyn514** [comment]
Line 1372 also needs to be part of the function or it will panic on that too:
```
         if path_str.contains(['<', '>'].as_slice()) {
            stripped_path_string = match strip_generics_from_path(path_str) {
```

File: src/librustdoc/passes/collect_intra_doc_links.rs [hunk 1]
```diff
@@ -892,6 +892,114 @@ impl<'a, 'tcx> DocFolder for LinkCollector<'a, 'tcx> {
...
+/// Returns:
+/// - `None` if the link should be ignored.
+/// - `Some(Err)` if the link should emit an error
+/// - `Some(Ok)` if the link is valid
+///
+/// `link_buffer` is needed for lifetime reasons; it will always be overwritten and the contents ignored.
+crate fn preprocess_link<'a>(
+    ori_link: &'a MarkdownLink,
+    link_buffer_1: &'a mut String,
+    link_buffer_2: &'a mut String,
```
Comment (jyn514) [inline src/librustdoc/passes/collect_intra_doc_links.rs:923]: `link_buffer` is probably overkill, happy to remove it if you think just copying the string is better.

**Comment by jyn514** [comment]
Ok, this should be ready for review.

File: src/librustdoc/passes/collect_intra_doc_links.rs (hunk 1 above)
Comment (camelid) [reply src/librustdoc/passes/collect_intra_doc_links.rs:923]: I find it a bit hard to follow the code with `link_buffer`, so I think it might be good to remove it for now. We already introduce copying with things like ``ori_link.link.replace("`", "")``, so it probably isn't that significant of a perf saving.

File: src/librustdoc/passes/collect_intra_doc_links.rs [hunk 2]
```diff
@@ -892,6 +892,114 @@ impl<'a, 'tcx> DocFolder for LinkCollector<'a, 'tcx> {
...
+) -> Option<Result<PreprocessingInfo<'a>, PreprocessingError<'a>>> {
+    // Bail early for real links.
+    if ori_link.link.contains('/') {
+        return None;
+    }
+
+    // [] is mostly likely not supposed to be a link
+    if ori_link.link.is_empty() {
+        return None;
+    }
```
Comment (klensy) [inline src/librustdoc/passes/collect_intra_doc_links.rs:933]: This can be swapped.

File: src/librustdoc/passes/collect_intra_doc_links.rs [hunk 3]
```diff
@@ -892,6 +892,114 @@ impl<'a, 'tcx> DocFolder for LinkCollector<'a, 'tcx> {
...
+                // Only report error if we would not have ignored this link.
+                // See issue #83859.
+                let disambiguator_range = (no_backticks_range.start + relative_range.start)
+                    ..(no_backticks_range.start + relative_range.end);
+                return Some(Err(PreprocessingError::Disambiguator(disambiguator_range, err_msg)));
+            } else {
+                return None;
+            }
+        }
+    };
```
Comment (klensy) [inline src/librustdoc/passes/collect_intra_doc_links.rs:965]: `no_backticks_range` can be moved inside `if !should_ignore_link_with_disambiguators(link)`

File: src/librustdoc/passes/collect_intra_doc_links.rs (hunk 1 above)
Comment (jyn514) [reply src/librustdoc/passes/collect_intra_doc_links.rs:923]: Hmm, I'm not quite show how - I tried adding `to_owned()` in a few places but I still got an error:
```

error[E0515]: cannot return value referencing temporary value
   --> src/librustdoc/passes/collect_intra_doc_links.rs:981:24
    |
935 |       let mut parts = ori_link.link.replace("`", "").split('#');
    |                       ------------------------------ temporary value created here
...
981 |                   return Some(Err(PreprocessingError::Resolution(
    |  ________________________^
982 | |                     err_kind,
983 | |                     path_str.into(),
984 | |                     disambiguator,
985 | |                 )));
    | |___________________^ returns a value referencing data owned by the current function
```
The problem is `path_str` is borrowed from `link`, which is a local `String`, and so I can't return it.

Comment (jyn514) [reply src/librustdoc/passes/collect_intra_doc_links.rs:923]: Oh never mind I figured it out, `into()` didn't do what I thought.

**Review by Manishearth — APPROVED** [review]

File: src/librustdoc/passes/collect_intra_doc_links.rs [hunk 4]
```diff
@@ -892,6 +892,117 @@ impl<'a, 'tcx> DocFolder for LinkCollector<'a, 'tcx> {
     }
 }
 
+crate enum PreprocessingError<'a> {
+    Anchor(AnchorFailure),
+    Disambiguator(Range<usize>, String),
```
Comment (Manishearth) [inline src/librustdoc/passes/collect_intra_doc_links.rs:897]: nit: worth having comments on each one explaining what the error is

**Comment by Manishearth** [comment]
@bors r-

I'm going to let @jyn514 figure out how to land this PR and https://github.com/rust-lang/rust/pull/84101

**Comment by camelid** [comment]
Hmm, somehow this PR still got merged after being `r-`'d?

**Comment by jyn514** [comment]
@camelid it was merged as part of https://github.com/rust-lang/rust/pull/84101

**Comment by camelid** [comment]
Ah, that makes sense.
